nautobot/nautobot_data_validation_engine/custom_validators.py

```diff
@@ -0,0 +1,309 @@
+                        }
+                    )
+
+        # DataComplianceRules
+        for compliance_class in get_data_compliance_rules_map().get(self.model, []):
+            compliance_class(obj).clean()
+
+        for repo in GitRepository.objects.filter(
+            provided_contents__contains="nautobot.nautobot_data_validation_engine.data_compliance_rules"  # TODO(john): data migration for this? used to be data_validation_engine.data_compliance_rules
+        ):
+            for compliance_class in get_classes_from_git_repo(repo):
+                if (
+                    f"{self.context['object']._meta.app_label}.{self.context['object']._meta.model_name}"
+                    != compliance_class.model
+                ):
+                    continue
+                compliance_class(self.context["object"]).clean()
+
+    def get_compliance_result(self, message=None, instance=None, attribute=None, valid=True):
+        """Generate a DataCompliance object based on the given parameters."""
+        attribute_value = getattr(instance, attribute, None)
+        class_name = f"{instance._meta.app_label.capitalize()}{instance._meta.model_name.capitalize()}CustomValidator"
+
+        result, _ = DataCompliance.objects.update_or_create(
+            compliance_class_name=class_name,
+            content_type=ContentType.objects.get_for_model(instance),
+            object_id=instance.id,
+            validated_attribute=attribute,
+            defaults={
+                "last_validation_date": timezone.now(),
+                "validated_object_str": str(instance),
+                "validated_attribute_value": str(attribute_value) if attribute_value else "",
+                "message": message,
+                "valid": valid,
+            },
+        )
+
+        return result
+
+
+def is_data_compliance_rule(obj):
+    """Check to see if object is a DataComplianceRule class instance."""
+    return inspect.isclass(obj) and issubclass(obj, DataComplianceRule) and obj is not DataComplianceRule
+
+
+def get_data_compliance_rules_map():
+    """Generate a dictionary of audit rulesets associated to their models."""
+    compliance_rulesets = {}
+    for validators in registry["plugin_custom_validators"].values():
+        for validator in validators:
+            if is_data_compliance_rule(validator):
+                compliance_rulesets.setdefault(validator.model, [])
+                compliance_rulesets[validator.model].append(validator)
+
+    return compliance_rulesets
+
+
+def get_classes_from_git_repo(repo: GitRepository):
+    """Get list of DataComplianceRule classes found within the custom_validators folder of the given repo."""
+    ensure_git_repository(repo, head=repo.current_head)
+    class_list = []
+    for importer, discovered_module_name, _ in pkgutil.iter_modules([f"{repo.filesystem_path}/custom_validators"]):
```

Comment:
Do we have to reimport the code on *every* model clean? Worried about performance impact. At the very least, `ensure_git_repository()` will return `False` if the local repo clone was already up-to-date - can we skip the reloads in that case?